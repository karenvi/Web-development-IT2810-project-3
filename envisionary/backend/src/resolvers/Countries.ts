
import Country from "../models/models";


const resolvers = {
    Query: {
        getCountryByName(parents, args, context, info) {
            return Country.find({"Country" : "Norway"})
                .then(country => {
                    return country.map(result => ({ ...result }))
                })
                .catch(err => {
                    console.error(err);
                })
        }
    }
};