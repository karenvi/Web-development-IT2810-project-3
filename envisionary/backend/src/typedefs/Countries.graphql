type Country {
    Rank: String
    CCA3: String
    Country: String
    Capital: String
    Continent: String
    2022-Population: String
    2020-Population: String
    2015-Population: String
    2010-Population: String
    2000-Population: String
    1990-Population: String
    1980-Population: String
    1970-Population: String
    Area: String
    Density: String
    Growth-Rate: String
    World-Population-Percentage: String
}