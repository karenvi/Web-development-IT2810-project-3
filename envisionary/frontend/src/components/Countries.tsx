import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import { useNavigate } from 'react-router-dom';
import { useQuery, gql } from '@apollo/client';
import { useRecoilState } from 'recoil';
import { categoryState, searchQueryState } from '../states/states';

function Countries() {
  const [searchQuery, setSearchQuery] = useRecoilState(searchQueryState);
  // const [countries, setCountries] = useState<ICountry[]>([]);
  const navigate = useNavigate()
  const [category, setCategory] = useRecoilState(categoryState);

  // Code for the searchbar
  const { search } = window.location;
  const query = new URLSearchParams(search).get('s');
  if (query) {
    setSearchQuery(query);
  }
  // Code for displaying data from the database
  const GET_COUNTRIES = gql`
    query Countries {
      countries {
        Rank,
        CCA3,
        Country,
        Capital,
        Continent,
        Population2022,
        Population2020,
        Population2015,
        Population2010,
        Population2000,
        Population1990,
        Population1980,
        Population1970,
        Area,
        Density,
        GrowthRate,
        WorldPopulationPercentage
      }
    }
  `;

  const { loading, error, data } = useQuery(GET_COUNTRIES);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  interface ICountry {
    _id: {
      $oid: string
    },
    Rank: string,
    CCA3: string,
    Country: string,
    Capital: string,
    Continent: string,
    Population2022: string,
    Population2020: string,
    Population2015: string,
    Population2010: string,
    Population2000: string,
    Population1990: string,
    Population1980: string,
    Population1970: string,
    Area: string,
    Density: string,
    GrowthRate: string,
    WorldPopulationPercentage: string
  }

  // const fetchData = () => {
  //   setCountries(countriesJson)
  // }

  const toCountryPage = (country: ICountry) => {
    navigate('/country', {state: {country}})
  }

  const tableHeadStyling = {fontWeight: 'bold'}

  const filterCountries = (countries: any, query: String | null) => {
    if (!query) {
      return countries;
    } 

    // Make sure query is valid even though user types it without big capital letters etc.
    if (query) {
      const lengthOfQuery = query.length;
      const modifiedQuery = query[0].toUpperCase() + query.substring(1, lengthOfQuery).toLowerCase();
      query = modifiedQuery;
    } 

    // TODO: Må finne ut hvordan man kan bare bruke category i stedet for å hardkode inn alle verdiene :,)
    return countries.filter((country: any) => {
      if (category == "Country") {
        const countryName = country.Country;
        if (countryName != null) {
          if (countryName.includes(query)) {
            return countryName.includes(query)
          }
        }
      } else if (category == "Population") {
        const countryPopulation = country.Population2022;
        if (countryPopulation != null) {
          if (countryPopulation.includes(query)) {
            return countryPopulation.includes(query)
          }
        }
      } else if (category == "Area") {
        const countryArea = country.Area;
        if (countryArea != null) {
          if (countryArea.includes(query)) {
            return countryArea.includes(query);
          }
        }
      } else if (category == "Continent") {
        const countryContinent = country.Continent;
        if (countryContinent != null) {
          if (countryContinent.includes(query)) {
            return countryContinent.includes(query);
          }
        }
      }
      
    })
  }

  const testCountries = filterCountries(data.countries, searchQuery)
  console.log(testCountries)

  return (
    <TableContainer sx={{ width: '50%', m: '10px' }} component={Paper}>
      <Table sx={{ minWidth: 300 }} aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell sx={tableHeadStyling}>Country</TableCell>
            <TableCell sx={tableHeadStyling} align="right">Continent</TableCell>
            <TableCell sx={tableHeadStyling} align="right">Population (2022)</TableCell>
            <TableCell sx={tableHeadStyling} align="right">Area (km&#178;)</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {testCountries.map((row: any ) => (
            <TableRow
              sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
              onClick={() => {toCountryPage(row)}}
              hover={true}
            >
              <TableCell component="th" scope="row">{row.Country}</TableCell>
              <TableCell align="right">{row.Continent}</TableCell>
              <TableCell align="right">{parseInt(row.Population2022).toLocaleString()}</TableCell>
              <TableCell align="right">{parseInt(row.Area).toLocaleString()}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}
export default Countries