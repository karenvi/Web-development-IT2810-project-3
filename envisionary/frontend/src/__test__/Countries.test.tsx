import React from 'react';
import Countries from '../components/Countries';
export {};



it("test filterCountries", () => {
    
});