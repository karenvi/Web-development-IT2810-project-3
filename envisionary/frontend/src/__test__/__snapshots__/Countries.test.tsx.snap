// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Countries component snapshot test 1`] = `
<main
  className="MuiBox-root css-hbm2t4"
>
  <div
    className="MuiBox-root css-eolfkn"
  >
    <form
      action="/"
      autoComplete="off"
      method="get"
      onSubmit={[Function]}
    >
      <label
        htmlFor="header-search"
      >
        <span
          className="visually-hidden"
        >
          Search by 
          country
        </span>
      </label>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        onInput={[Function]}
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="header-search"
          id="header-search-label"
        >
          Search for country
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="search-test"
            disabled={false}
            id="header-search"
            name="s"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Search for country
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </form>
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth css-1tp1e4i-MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        id="demo-simple-select-label"
      >
        Category:
      </label>
      <label
        htmlFor="header-search"
      >
        <span
          className="visually-hidden"
        >
          Select what to search category to search in
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="demo-simple-select-label demo-simple-select"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="demo-simple-select"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Country
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          data-testid="select-test"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="Country"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-14lo706"
          >
            <span>
              Category:
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-6ilaaq-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="Table of countries"
      className="MuiTable-root css-rqglhn-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-11sptls-MuiTableCell-root"
            colSpan={2}
            scope="col"
          >
            <label
              htmlFor="header-search"
            >
              <span
                className="visually-hidden"
              >
                Sort by:
              </span>
            </label>
            <div
              className="MuiFormControl-root MuiFormControl-fullWidth css-im9t9x-MuiFormControl-root"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink={true}
                id="select-filter-category"
              >
                Sort by:
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                onClick={[Function]}
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="select-filter-category filter-category"
                  className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="filter-category"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Ascending country
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="Country-asc"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden={true}
                  className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="css-14lo706"
                  >
                    <span>
                      Sort by:
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
            colSpan={2}
            scope="col"
          >
            Hide unreviewed countries
            <span
              className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <input
                aria-label="controlled"
                checked={false}
                className="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate={false}
                onChange={[Function]}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </th>
        </tr>
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-11sptls-MuiTableCell-root"
            scope="col"
          >
            Country
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
            scope="col"
          >
            Continent
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
            scope="col"
          >
            Population (2022)
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1o98s2t-MuiTableCell-root"
            scope="col"
          >
            Area (km²)
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colSpan={4}
          >
            <div
              className="MuiGrid-root MuiGrid-container css-zo8441-MuiGrid-root"
            >
              <div
                className="MuiGrid-root css-gfmvt9-MuiGrid-root"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1lf956w-MuiButtonBase-root-MuiButton-root"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  type="button"
                >
                  Previous
                </button>
              </div>
              <div
                className="MuiGrid-root css-1yi6br4-MuiGrid-root"
              >
                Page 
                1
              </div>
              <div
                className="MuiGrid-root css-1ivdjg1-MuiGrid-root"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1lf956w-MuiButtonBase-root-MuiButton-root"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  Next
                </button>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</main>
`;
